geode-gfsh/src/test/java/org/apache/geode/management/internal/cli/commands/StopGatewaySenderCommandTest.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.apache.geode.management.internal.cli.commands;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doReturn;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Test;
+
+import org.apache.geode.distributed.DistributedMember;
+import org.apache.geode.management.internal.SystemManagementService;
+import org.apache.geode.management.internal.cli.result.model.ResultModel;
+import org.apache.geode.management.internal.functions.CliFunctionResult;
+import org.apache.geode.test.junit.rules.GfshParserRule;
+
+public class StopGatewaySenderCommandTest {
+  @ClassRule
+  public static GfshParserRule gfsh = new GfshParserRule();
+
+  private StopGatewaySenderCommand command;
+
+  @Before
+  public void before() {
+    command = spy(StopGatewaySenderCommand.class);
+    List<CliFunctionResult> functionResults = new ArrayList<>();
+    doReturn(functionResults).when(command).executeAndGetFunctionResult(any(), any(), any());
```

Comment:
Anytime you override an instance method on the component being tested, you're either not testing some part of the component or it suggests that the method represents a responsibility that should not be on this component.

You can avoid using `spy` and overriding the instance method by using a field of type `java.util.function.Function`, or a custom interface, that is injected into the constructor(s).

1. Add cleanResults field of type `java.util.function.Function` or custom inner-interface
```
private final java.util.function.Function<List<?>, List<CliFunctionResult>> cleanResults;
```
2. Add constructors to `GfshCommand`
```
  public GfshCommand() {
    this(CliFunctionResult::cleanResults);
  }

  protected GfshCommand(java.util.function.Function<List<?>, List<CliFunctionResult>> cleanResults) {
    this.cleanResults = cleanResults;
  }
```
3. Add constructors to `StopGatewaySenderCommand`
```
  @SuppressWarnings("unused") // invoked by spring shell
  public StopGatewaySenderCommand() {
    this(CliFunctionResult::cleanResults);
  }

  StopGatewaySenderCommand(Function<List<?>, List<CliFunctionResult>> cleanResults) {
    super(cleanResults);
  }
```
4. Change `GfshCommand.executeAndGetFunctionResult` to use the field `cleanResults`
```
  public List<CliFunctionResult> executeAndGetFunctionResult(Function<?> function, Object args,
      Set<DistributedMember> targetMembers) {
    ResultCollector<?, ?> rc = executeFunction(function, args, targetMembers);
    return cleanResults.apply((List<?>) rc.getResult());
  }
```
5. Pass in simple mock when constructing `StopGatewaySenderCommand` in test
```
Function<List<?>, List<CliFunctionResult>> cleanResults = mock(Function.class);
command = new StopGatewaySenderCommand(cleanResults);
```
If you need to observe `cleanResults` to validate interaction with that dependency, you simply move it from a local variable to a field:
```
private Function<List<?>, List<CliFunctionResult>> cleanResults;
// ...
cleanResults = mock(Function.class);
command = new StopGatewaySenderCommand(cleanResults);
```
Then you can perform any Mockito validation calls against `cleanResults` that you need to.

Reply:
@kirklund Thanks a lot for the suggestion. I finally have not needed to modify `GfshCommand`.
Nevertheless, I had to inject the command delegate and the `findMembers` test behavior using the constructor.
The outcome has not been very clean. If you have any suggestions to make it cleaner, please let me know.

Given that `findMembers` is not a static method, I could not pass it when calling the last constructor. Instead, I had to set it after I called it.
On the other hand, the `StopGatewaySenderCommandDelegateParallelImpl` constructor needed the a `ManagementSystem` which is obtained by calling the instance `getManagementSystem()` method. As a result, I had to set it to null in the Spring shell constructor and later set it to the right value if it was found null.